Improve terminology provider error handling

csandersdev opened the pull request:
This changeset resolves cqframework/clinical_quality_language#929, cqframework/clinical_quality_language#921, cqframework/clinical_quality_language#920, and cqframework/clinical_quality_language#919 by introducing logic
to:

1) Strip the urn:oid prefix from ValueSetInfo.id if it is there.
2) Throw an UnsupportedOperationException if the CQL library tries to
use ValueSetInfo.version or ValueSetInfo.codeSystems.
3) Throw an IllegalArgumentException if the result of a URL-based query
returns an unexpected number of results (0 or >1)

Additionally, I updated the lookup operation to use named parameters vs.
positional indexes when dealing with the operation response which seemed
like a more reliable pattern.

As much as was reasonable I tried to mirror the logic in
hapi-fhir-jpaserver-cql/src/main/java/ca/uhn/fhir/cql/r4/provider/JpaTerminologyProvider.java.

I added a test dependency on wiremock and I use it to mock FHIR
interactions for test purposes.

csandersdev committed:
Improve terminology support error handling

This changeset resolves #437, #459, #460, and #461 by introducing logic
to:

1) Strip the urn:oid prefix from ValueSetInfo.id if it is there.
2) Throw an UnsupportedOperationException if the CQL library tries to
use ValueSetInfo.version or ValueSetInfo.codeSystems.
3) Throw an IllegalArgumentException if the result of a URL-based query
returns an unexpected number of results (0 or >1)

Additionally, I updated the lookup operation to use named parameters vs.
positional indexes when dealing with the operation response which seemed
like a more reliable pattern.

As much as was reasonable I tried to mirror the logic in
hapi-fhir-jpaserver-cql/src/main/java/ca/uhn/fhir/cql/r4/provider/JpaTerminologyProvider.java.

I added a test dependency on wiremock and I use it to mock FHIR
interactions for test purposes.

csandersdev committed:
Fix a small bug in parameter creation for the lookup test.

JPercival commented:
Thanks for the contribution and in particular thanks for creating a pattern for unit tests of the TerminologyProviders. Going to take a look at this more in depth today.

csandersdev committed:
Fix ID overwrite to not cause duplicate ValueSet prefix

csandersdev commented:
Thanks, @JPercival. Let me know if there is anything additional I can do here. The Travis CI build is reporting as failed and I can't tell why without logs. Works for me locally. Also, I PR'ed to main because there isn't a current develop branch. Not sure if you would want to create dev branch and merge there first to queue this up for a future release or just go direct.

csandersdev committed:
Respond to code review feedback from my team.

* The WireMock http port was not being assigned using the dynamic
assignment feature that was built into the test because it defaulted to
8080 in the static initializer.
* There was an unnecessary and somewhat error prone check around the
unsupported operation exception in resolveByURL that was copied from the
HAPI JPA implementation
* The Bundle.isEmpty() method does not behave with the same semantics as
checking to Bundle.getEntry().isEmpty() and is potentially error prone.
* The display value in lookup would potentially overwrite a provided
display value with null when the parameter was not provided in the
lookup response. Defaulting to no touch now when it isn't there.

JPercival commented:
The code path to resolve by ValueSets by Id was actually a workaround for some cases where ValueSets didn't have a url. Specifically, having the oid as the identifier is common in QDM-land. The the VSAC terminology is also non-FHIR compliant in this regard. Given the guidance from from the [CQL specification](https://cql.hl7.org/02-authorsguide.html#terminology)

> The external identifier need not be an OID, it may be a uniform resource identifier (URI), or any other identification system. CQL does not interpret the external id, it only specifies that the external identifier be a string that can be used to uniquely identify the valueset within the implementation environment.

And the guidance from the [FHIR specification](https://www.hl7.org/fhir/resource.html#canonical)

> The canonical URL is the preferred way to reference a resource instance for the resource types on which it is defined.
> Note that the element actually contains a URI, but is named url for legacy reasons.

And the fact that urn:uuid / urn:oid are valid uris, I think that the urn prefix should probably first be resolved as a url rather than an id. IOW, I think the resolution should be:

1. url
2. identifier
3. id, having the urn prefix stripped

A hypothetical VSAC ValueSet would look something like:

```json
{
"url" : "vsac.com/1.2.3.4",
"identifier" : "urn:oid:1.2.3.4",
"id": "1.2.3.4" 
}
```

Kind of a pain, but not sure how else to square the spec with what's out in the wild. Also makes caching the terminology that's fetched more important due to the likelihood of misses. There's a naive implementation of that over here:

https://github.com/DBCG/cql-evaluator/blob/master/evaluator.engine/src/main/java/org/opencds/cqf/cql/evaluator/engine/terminology/PrivateCachingTerminologyProviderDecorator.java

csandersdev commented:
For my use case, we enforce the URL requirement, so urn:oid: isn't an issue. I got down that path because the CQL Author's Guide has quite a few examples that use the urn:oid:XXX pattern and those examples were failing in the TerminologyProvider expand method.  To your point, though, I get the need to handle multiple lookup paths and I'm happy to make the suggested change.

In my testing I've also noticed that the expand operation is sending a POST with the body ``{ "resourceType": "Parameters" }`` and that is not handled well by my IBM FHIR server. I would like to switch that request to a GET if there aren't any objections. I tested it against http://hapi.fhir.org/baseR4 and it works fine both there and on my local IBM FHIR.

JPercival commented:
The DataProviders based on the FHIR Rest API have configuration that allows you to choose a GET or a POST depending on your specific server. I wouldn't be opposed to adding similar configuration for the TerminologyProviders. There are other servers besides hapi that are commonly used in cql-engine deployments, such as Apelon, so I don't think we should remove POST outright.

csandersdev commented:
POST would make sense if the expand operation were taking advantage of the input parameters, such as if it tried to do ``/ValueSet/$expand?url=myurlhere``, but right now the resource resolution is done in a separate call and always requests ``/ValueSet/myidhere/$expand`` with no input parameters.  That is a POST right now with an empty parameters object and that POST is what is causing the problem.

I checked around with our local FHIR experts and the consensus is that the expand operation does not affect state and can be converted to a GET as described in the spec [here](https://www.hl7.org/fhir/operations.html#executing) even though the operation definitions seem lax in advertising it. They also mentioned that empty objects were not allowed in the spec as discussed [here](https://chat.fhir.org/#narrow/stream/179202-terminology/topic/CodeSystem.20.24lookup/near/190328714) until the vote was approved for R4B as described [here]. It seems safe to make this call a GET right now. I would be fully onboard supporting user-driven conversion between GET and POST at a later date as needed, but I think that is slightly out of scope for this PR. You could probably convince me otherwise, but I don't see a strong need to introduce that feature here.

JPercival commented:
I checked the history of the terminology provider to ensure there wasn't a reason we switched to POST. Couldn't find one, so I'm good with GET.

csandersdev committed:
Update resolveByUrl search path, switch expand to GET

Per feedback, I updated the resolveByUrl method to use a search path
that looks first for Resource.url, then Resource.identifier, and, if all
else fails *and* the value looks like a FHIR resource ID (stripping
urn:oid: or urn:uuid: prefixes) it tries to read by Resource.id.

This changeset also switches the expand operation to useHttpGet() vs.
the default POST operation because there are problems with POST'ing an
empty parameters object.

Please excuse the whitespace changes in the test class. My editor was
using tabs and the project seems to prefer spaces, so I updated the file
to be consistent with the rest of the project.

JPercival reviewed (approved):
Looks great. I'm going to pull this into a couple downstream projects to test and then I'll merge it.
